### PR TITLE
3D click + drag interactivity API

### PR DESCRIPTION
**Pull request by alisterburt**
# Description
This PR adds the machinery required for custom click and drag interactivity, separated out from #3152 so the API around this machinery can be discussed in isolation.

This adds `Layer.projected_distance_from_mouse_events(start_event, end_event, vectors)`, a method which calculates, in data coordinates, the projection of the drag vector between two mouse events onto a set of vectors defined in data coordinates.

As an example, if we define a *pseudocanvas*, the vispy canvas defined in the data coordinate system, and calculate the projection of the drag vector onto two orthogonal vectors on that pseudocanvas, we can move points around in the plane parallel to the canvas.

https://user-images.githubusercontent.com/7307488/130124723-fe4a91f7-848b-4681-bc9c-90f3d0b329be.mp4

The example code is a bit messy, mainly due to the inability to serialise events properly (issue at #3177), but should give you an idea of how these interactions can be implemented.

Main questions are to do with the API, there was discussion about whether we should use mouse events directly or their attributes in #3117, using the events directly feels more messy (layers know about mouse events) but puts more on the end user

### CODE REVIEW

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py]: I find this name and docstring hard to parse. Can you clarify a bit what this means?

File: napari/layers/base/base.py
Comment (brisvag) [inline napari/layers/base/base.py]: This is also a bit confusing; `display` to me sounds like a verb, so I expect this to display something, not return an array?

**Review by brisvag — COMMENTED** [review]
This looks super cool! Looking forward to this being merged :) I have a couple comments about naming, but the rest LGTM.

File: napari/layers/base/base.py
Comment (alisterburt) [reply napari/layers/base/base.py]: agree it's not the easiest to parse but not immediately sure of a better name... will improve the docstring

I'm not sure if this will still be removed as layer data sources are separated from layers but will leave until that happens 👍🏼 
```suggestion
    @property
    def _dims_displayed_mask(self):
        """**to be removed** displayed dimensions mask"""
```

File: napari/layers/base/base.py
Comment (alisterburt) [reply napari/layers/base/base.py]: agree - do you like `_bounding_box_displayed`?

**Review by alisterburt — COMMENTED** [review]
looking back at this - I will refactor the `mouse_events_to_projected_distance` function so that it takes start and end positions directly before asking for full review

**Comment by alisterburt** [comment]
okay I think this is ready for a first pass @brisvag - cc @kevinyamauchi if you have some time it would be great to get some feedback!

The concept of an nD view direction on the event still confuses me -> the view direction on the event feels like it should be the camera view direction at all times, can you convince me otherwise @kevinyamauchi ?

edit: rebased, hope that fixes tests but there may still be an issue

**Comment by alisterburt** [comment]
okay, test failures are now unrelated and this functionality has basic tests both on the utility function and on the layer. @brisvag @kevinyamauchi I think this is ready to go 🙂

File: napari/layers/utils/interactivity_utils.py
Comment (kevinyamauchi) [inline napari/layers/utils/interactivity_utils.py]: Are these lines to account for when the user passes something that isn't a numpy array (e.g., a list)? If so, it might make sense to use `asarray` instead to avoid unnecessary copying.

File: napari/layers/utils/interactivity_utils.py
Comment (kevinyamauchi) [inline napari/layers/utils/interactivity_utils.py]: Is this line required?

File: examples/interactive_move_point_3d.py
Comment (kevinyamauchi) [inline examples/interactive_move_point_3d.py:41]: I like this example! I am wondering if it is more complex than it needs to be. Rather than defining a new basis and then calculate the distances along those vectors, can you project `start_point_data` and `end_point_data` onto a plane parallel to the canvas and then add the vector `end_point_canv - start_point_canv` to the data?

Comment (kevinyamauchi) [reply examples/interactive_move_point_3d.py:41]: I wonder if dragging a plane (rectangle in the shapes layer) along it's normal would be a better example?

**Review by kevinyamauchi — COMMENTED** [review]
Thanks @alisterburt ! The test and function look good to me. I made a suggestion about the example and a couple of my minor questions about a couple of the variables below

File: napari/layers/utils/interactivity_utils.py
Comment (alisterburt) [reply napari/layers/utils/interactivity_utils.py]:
```suggestion
    start_position = np.asarray(start_position)
    end_position = np.asarray(end_position)
```

Comment (alisterburt) [reply napari/layers/utils/interactivity_utils.py]: good shout!

File: examples/interactive_move_point_3d.py
Comment (alisterburt) [reply examples/interactive_move_point_3d.py:41]: This is a good point - I've definitely overcomplicated it, will make the moving a plane example tomorrow

**Comment by alisterburt** [comment]
@kevinyamauchi I tidied up the example as you suggested (much better now!) and added a 'move rectangle' example

unfortunately modifying shapes data in place is a bit more complicated than I thought. Updating `Shapes.data` manually errored and shifts are only implemented in private APIs so for now I just add new rectangles... if you see an easy path to getting that sorted that doesn't dive into the private APIs I'm all ears!


https://user-images.githubusercontent.com/7307488/138968681-4b94f49f-8a93-431d-aa58-1b229439d1c8.mp4

File: napari/layers/utils/interactivity_utils.py
Comment (alisterburt) [reply napari/layers/utils/interactivity_utils.py]: nope, was just trying to be clear about coordinate systems but agree it's redundant, commit the proposed change below if you agree

File: napari/layers/utils/interactivity_utils.py
Comment (alisterburt) [inline napari/layers/utils/interactivity_utils.py]:
```suggestion
    end_position_canvas = project_point_onto_plane(
        end_position, start_position, view_direction
```

File: napari/layers/utils/interactivity_utils.py
Comment (alisterburt) [inline napari/layers/utils/interactivity_utils.py]:
```suggestion
        end_position_canvas - start_position
```

**Comment by brisvag** [comment]
Sweet example! I get some old windows "dragging-windows" vibes, but in 3D :D

Looking forward to this being merged. I got a bit lost in the changes: what's the current relation of this regarding rotations (if any of this is usable for that)?

**Comment by alisterburt** [comment]
hahah yeah I had the same feeling!

For 3D rotations you don't actually need this so much, you would probably do this by
1. calculating the drag vector in canvas coordinates
2. rotate that drag vector 90 degrees in canvas coordinates to define axis of rotation (still in canvas coordinates)
3. transform axis of rotation from canvas coordinates to data coordinates
3. rotate object around its center, along the axis of rotation

You might have to perform step 2 in data coordinates with the transforms you have exposed, let me know if this doesn't make sense

File: napari/layers/base/base.py
Comment (sofroniewn) [inline napari/layers/base/base.py]: Can we make this private?

File: napari/layers/base/base.py
Comment (sofroniewn) [inline napari/layers/base/base.py:1091]: I'm ok with this name, it's a little long - but makes sense. Also makes sense to be public

**Review by sofroniewn — APPROVED** [review]
LGTM! Very happy to have this go in with the nice example. I think with time we'll want to think about how to add access to this functionality in the GUI alone.

Maybe now is the time also to start writing up a 3D interactivity guide as part of our docs. We could highlight the different examples you have now and maybe also the new additions to the API including the experimental ones. Thoughts @alisterburt @kevinyamauchi?

**Comment by kevinyamauchi** [comment]
> Maybe now is the time also to start writing up a 3D interactivity guide as part of our docs. We could highlight the different examples you have now and maybe also the new additions to the API including the experimental ones. Thoughts @alisterburt @kevinyamauchi?

@sofroniewn , we have some 3D interactivity docs [here](https://napari.org/guides/stable/3D_interactivity.html). Is this along the lines of what you had in mind? After this merges, we can add this new functionality there!

**Comment by sofroniewn** [comment]
> @sofroniewn , we have some 3D interactivity docs here. Is this along the lines of what you had in mind? After this merges, we can add this new functionality there!

Ooh right yes! (i'd forgotten, my bad) Let's keep updating, i think the experimental API might need mentioning there too. Also it might be fun to include links to videos for those examples too 🚀

**Review by kevinyamauchi — APPROVED** [review]
Thanks, @alisterburt ! This looks good to me! I think there's a merge conflict that needs to be resolved, but if the tests pass after that, I think this is good to merge.

File: napari/layers/base/base.py
Comment (kevinyamauchi) [reply napari/layers/base/base.py]: I agree with making this private.

**Comment by kevinyamauchi** [comment]
> Ooh right yes! (i'd forgotten, my bad) Let's keep updating, i think the experimental API might need mentioning there too. Also it might be fun to include links to videos for those examples too 🚀

All good! Good point about mentioning the experimental API. In general, it would be great to have more videos on that page to better highlight how the interaction works/looks!

**Comment by alisterburt** [comment]
Thank you both for taking the time to review @kevinyamauchi and @sofroniewn - I have integrated all suggested changes and tidied up a few small bits. Will merge once tests pass! 🎉 

edit: merged, also agree that docs will be super important here 👍🏼

**Pull request merged**
